### DISCUSSION TITLE
Remove edited/deleted files from library

### DISCUSSION THREAD

**Discussion by accollie** [music-assistant/support#2159] [Feature requests and ideas]
I think it would be cool to remove missing or edited files from the library on sync. For example if i change a file on my server or delete it. When i sync it should be removed from the music assistant library.

**Comment by OzGav** [comment]
This should already be happening

**Reply by accollie** [reply]
So I'm in the artist tab. Im trying to clean up my library by deleting some artists on my server. The song is removed but the artist listing remains.

**Reply by OzGav** [reply]
Select the three dot menu in the banner at the top and remove from library

**Reply by accollie** [reply]
Sorry OsGav,
I meant that it would be nice for the library element to be updated/deleted upon a manual sync/refresh. Or even restart of the addon. I have been manually deleting the folders from the user interface using the three dots. I love the program and the work that has been put into this project. This suggestion is just for a little polish. It is something that i have become accustomed to using Plex.

**Reply by marcelveldt** [reply]
it already works like this

**Reply by OzGav** [reply]
From Marcel it should already work like that but it does do a double check if no other providers are linked. so if no other providers are linked anymore, the item gets cleaned up

**Reply by accollie** [reply]
Maybe im doing something wrong. I delete the file from my file share which is on a network storage to homeassistant. I the go to file system storage and the three dots (synchronize now). When i return to the Artists Tab the artist name is still listed (the song file is gone however). In my opinion, if there are no tracks linked to the artist it should also not exist. Ive been playing around with this now for a few hours. It takes forever to go through and delete the one by one.

**Reply by OzGav** [reply]
Do you have any other music providers added?

**Reply by accollie** [reply]
Just Filesystem (local disk). Im starting to think that is the problem. I will try using Filesystem (remote share) to connect to my OMV Folder. At the moment the folder is mounted to HA.

**Reply by OzGav** [reply]
Not necessarily. Might be a bug. Will take a few days for me to try and reproduce as I am away from home

**Reply by OzGav** [reply]
Ok looking at the behaviour with the current beta, as soon as I removed the only track I had for a particular artist the tracks and albums from my steaming providers were then displayed as options to add to the library. So MA would not automatically remove the artist in this case. Do you have any streaming providers?

**Reply by accollie** [reply]
No i don't have any streaming services. Im using Open Media Vault.

**Reply by OzGav** [reply]
ok I believe I have recreated this. I will transfer this to an issue.

**Comment by lindsaymarkward** [comment]
I know this is an old discussion, but I have the same problem, or similar.
I'm cleaning up my local files library, using Picard to tag "[unknown]" files and deleting some others.
MA still shows me these deleted files in the library, and I have the [unknown] tracks as well as the newly-tagged ones... even though I've pressed the sync button in the settings for the local disk provider and I've restarted the computer.
What am I missing, and what can I do?
If this was transferred to an issue, as @OzGav said, which issue # is it so I can check on the state of this?
Thanks!

**Reply by OzGav** [reply]
Just select the track you want to get rid of and REMOVE FROM LIBRRY

**Reply by lindsaymarkward** [reply]
While I understand that method...
a. I have deleted and changed very many songs/files. I have no easy way of knowing which ones need to be deleted/changed and I'd rather not do this manually.
b. Is this really the normal way to do it? What does 'sync' mean if syncing does not update the library by removing missing files and updating metadata?
I thought the comment above:

> This should already be happening

meant that the sync was supposed to do what I'm expecting. 
Is there an issue for this?

**Reply by OzGav** [reply]
Possibly depends on what provider you’re using and if you have multiple providers. It is possible that many unrelated changes in and outside the library can break the sync mechanism. If all else fails you can always reset the database and get it rebuilt from scratch.

**Reply by lindsaymarkward** [reply]
Thanks so much for the replies. I'm still getting used to MA... 
I referred to "local files library" so I thought that means it must be the  "Filesystem (local disk)" provider. (Is it possible to have a different provider for local files?)
I do have a few other providers like Spotify, TuneIn, but I'm not sure if that's relevant.

I'm happy to reset the database, but I'm also interested in how it's supposed to work.
I searched the docs but can't find anything about how to reset the database. Is this described somewhere?

I take it there was no issue made for this earlier? I can't find one.

**Reply by OzGav** [reply]
There was an issue look at the top of this page on the right under EVENTS. 

We don’t advertise how to reset the db as ideally that shouldn’t be required. I assume you have restarted MA as that should re-trigger the cleanup but be patient as this is done very slowly. If you have restarting MA before the cleanup is complete that will mess things up.

You have two options from here. Remove the filesystem provider and then WAIT until the cleanup is complete. Look in the logs for this! If you are happy to go nuclear and can’t wait then you can go to MA SETTINGS >> CORE >> MUSIC CONTROLLER >> ADVANCED

Once you have tidied it up then we can check for an issue. You need to do one thing and the check if everything is cleaned up. If there is no clean up then report back what you have done step by step so we can try and recreate it.

**Reply by lindsaymarkward** [reply]
Thanks for pointing out "events". I hadn't seen this before. Perhaps I restarted MA (and/or the PC) before the cleanup was complete. Since you say that would "mess things up", I may be in that state now.
I've "delete"d the filesystem provider. Logs show:

> Removal of local provider detected, issuing full database reset...

I'll see how that goes, thanks.

**Reply by lindsaymarkward** [reply]
OK, so I removed the local filesystem provider and waited overnight. I then added it again. My number of "[unknown]" has now gone from 69 down to 47 and I can't (yet) find any of the songs I deleted.
So, that seemed to work.
I will continue to update my local files as needed, and see what happens when I sync the library, but I'll avoid restarting MA (if that was the issue, maybe).
Thanks.
